feat: Implement more advanced UI toolbar example

connorgmeehan opened the pull request:
# Objective

I would like an example of a more real world use case of the UI.  Showing / hiding elements contextually, more complexity in the hover states (selected elements having a different colour to unselected) and events to be consumed elsewhere in the app.

## Solution

In this instance I have added an example for a toolbar that tracks its internal state (used for styling and hiding the stop button if it's not currently playing/paused) and sends selection events that can be used by the rest of the game.

I am very new to rust so please critique my code styling / present better solutions / more idiomatic ways of doing things if you have the time.

Changes log is minimal, adds an example to the `examples/ui/` folder called `toolbar`.  Run with `cargo run --examples toolbar`

Cheers!

https://user-images.githubusercontent.com/7402063/195756902-f2159611-4d3b-4c5c-8a60-5fbd217abbec.mov

Some side notes: 
- I'm getting pretty poor performance on my laptop.  Even though I only have integrated graphics, click events have a pretty big delay, even on the `game_menu` example.  I'm on MacOS if it makes any difference, I wonder if this is something I could help address.  If so, let me know if I should open an issue for this because I'm not really sure what information is most helpful.
- It would be nice have a new utility function in `UiRect` to set the `left/right` and `top/bottom` values together as it's probably my most used way of setting padding / margin when doing web development.
I could PR something like this if interested but understand if you want to keep the framework light.
```rust
let rect = UiRect::from_x_y(Val::Px(10.0, Val::Px(5.0))
rect.left; // 10.0
rect.right; // 10.0
rect.top; // 5.0
rect.bottom; // 5.0
```

MrGVSV commented on examples/ui/toolbar.rs:
```diff
@@ -0,0 +1,279 @@
+/// Help wanted: I had to remove `Changed<Interaction>` from the query so that
+/// the buttons that weren't clicked would update.
+///
+/// Is there a way of listening to changes in `ResMut<State<ToolbarState>>`?
```
[`ResMut::is_changed`](http://dev-docs.bevyengine.org/bevy/ecs/change_detection/trait.DetectChanges.html#tymethod.is_changed) maybe?

connorgmeehan replied:
I forgot to remove this from my first commit where the entire state was managed in `ToolbarState` (including the message and the selected element) and there were no EventWriters/Readers so probably not worth changing now but thanks, that's exactly what I was looking for!

MrGVSV commented on examples/ui/toolbar.rs:
```diff
@@ -0,0 +1,279 @@
+const PRESSED_BUTTON: Color = Color::rgb(0.4, 0.4, 0.8);
+const SELECTED_BUTTON: Color = Color::rgb(0.35, 0.35, 0.75);
+
+/// System: Manages the click interaction and styling of all the button elements
```
Nit: You probably don't need to indicate this is a system since it's in the function name
```suggestion
/// Manages the click interaction and styling of all the button elements
```

MrGVSV commented on examples/ui/toolbar.rs:
```diff
@@ -0,0 +1,279 @@
+}
+
+#[derive(Component, Clone, Hash, Debug, Eq, PartialEq)]
+enum ButtonTags {
```
Nit: Enums are generally singular (i.e. `ButtonTag`)

MrGVSV commented on examples/ui/toolbar.rs:
```diff
@@ -0,0 +1,279 @@
+                    interaction: interaction.clone(),
+                });
+            }
+            Interaction::Hovered => {
+                ev_toolbar_button.send(ToolbarEvent {
+                    tag: button_tag.clone(),
+                    message: None,
+                    interaction: interaction.clone(),
+                });
+            }
+            Interaction::None => {
+                ev_toolbar_button.send(ToolbarEvent {
+                    tag: button_tag.clone(),
+                    message: None,
+                    interaction: interaction.clone(),
+                });
+            }
```
Nit: I think clippy will complain about this not being 
```suggestion
            Interaction::Hovered | Interaction::None => {
                ev_toolbar_button.send(ToolbarEvent {
                    tag: button_tag.clone(),
                    message: None,
                    interaction: interaction.clone(),
                });
            }
```

connorgmeehan replied:
Ahh I didn't know this was possible, thanks!

MrGVSV commented on examples/ui/toolbar.rs:
```diff
@@ -0,0 +1,279 @@
+            let clicked = this_button && ev.interaction == Interaction::Clicked;
+            let selected = tag.clone() == toolbar_state.current().selected;
+
+            // println!(
+            //     "Checking {:?} hovered?: {:?} clicked: {:?} selected: {:?}",
+            //     tag, hovered, clicked, selected
+            // );
```
Should this be removed?
```suggestion
```

MrGVSV commented on examples/ui/toolbar.rs:
```diff
@@ -0,0 +1,279 @@
+            //     tag, hovered, clicked, selected
+            // );
+
+            if selected {
+                if clicked {
+                    *color = PRESSED_BUTTON.into();
+                } else {
+                    *color = SELECTED_BUTTON.into();
+                }
+            } else {
+                if clicked {
+                    *color = PRESSED_BUTTON.into();
+                } else if hovered {
+                    *color = HOVERED_BUTTON.into();
+                } else {
+                    *color = NORMAL_BUTTON.into();
+                }
+            }
```
Nit: Could _maybe_ use a match statement
```suggestion
            match (selected, ev.interaction) {
                (_, Interaction::Clicked) => *color = PRESSED_BUTTON.into(),
                (true, _) => *color = SELECTED_BUTTON.into(),
                (_, Interaction::Hovered) => *color = HOVERED_BUTTON.into(),
                (_, _) => *color = NORMAL_BUTTON.into(),
            }
```

MrGVSV commented on examples/ui/toolbar.rs:
```diff
@@ -0,0 +1,279 @@
+                }
+            }
+
+            if tag.clone() == ButtonTags::Stop
```
Why clone? Couldn't you just do:
```suggestion
            if tag == &ButtonTags::Stop
```

MrGVSV commented on examples/ui/toolbar.rs:
```diff
@@ -0,0 +1,279 @@
+            let this_button = ev.tag.clone() == tag.clone();
+            let hovered = this_button && ev.interaction == Interaction::Hovered;
+            let clicked = this_button && ev.interaction == Interaction::Clicked;
+            let selected = tag.clone() == toolbar_state.current().selected;
```
You probably don't need to clone here either:
```suggestion
            let selected = tag == &toolbar_state.current().selected;
```

connorgmeehan replied:
Same as earlier, ran into issues comparing references, resolved by de-referencing.

MrGVSV commented on examples/ui/toolbar.rs:
```diff
@@ -0,0 +1,279 @@
+    for ev in ev_toolbar_button.iter() {
+        for (mut style, mut color, tag) in &mut style_query {
+            // Update button colours
+            let this_button = ev.tag.clone() == tag.clone();
```
Nit: You probably don't need to clone:
```suggestion
            let this_button = &ev.tag == tag;
```

connorgmeehan replied:
IIRC I ran into issues comparing references so I resolved this one by dereferencing as needed.

MrGVSV commented on examples/ui/toolbar.rs:
```diff
@@ -0,0 +1,279 @@
+
+/// System: Show and hide the toolbar status message when there is a message avaliable
+fn toolbar_message_system(
+    mut ev_toolbar_button: EventReader<ToolbarEvent>,
```
You could probably also just checked for a change in the toolbar's state instead of iterating events.

MrGVSV commented on examples/ui/toolbar.rs:
```diff
@@ -0,0 +1,279 @@
+//! This example shows a more advanced ui use case of managing a shared state via a set of button
+//! components.
+//! interaction state.
```
Is this meant to be here?
```suggestion
```

MrGVSV commented on examples/README.md:
```diff
@@ -311,6 +311,7 @@ Example | Description
 [Font Atlas Debug](../examples/ui/font_atlas_debug.rs) | Illustrates how FontAtlases are populated (used to optimize text rendering internally)
 [Text](../examples/ui/text.rs) | Illustrates creating and updating text
 [Text Debug](../examples/ui/text_debug.rs) | An example for debugging text layout
+[Toolbar](../examples/ui/toolbar.rs) | Showcases a more advanced UI with conditional rendering, events and managed state
```
Nit: I forget Bevy's stance on this (so maybe someone will correct me), but maybe add an Oxford comma? (Same for Cargo.toml)
```suggestion
[Toolbar](../examples/ui/toolbar.rs) | Showcases a more advanced UI with conditional rendering, events, and managed state
```

MrGVSV reviewed (commented):
I really like how clean and simple this example is. Good job!

MrGVSV commented:
> * I'm getting pretty poor performance on my laptop.  Even though I only have integrated graphics, click events have a pretty big delay, even on the `game_menu` example.  I'm on MacOS if it makes any difference, I wonder if this is something I could help address.  If so, let me know if I should open an issue for this because I'm not really sure what information is most helpful.

I'm on Mac as well (M1 Max) and it runs fine for me. Although, I did notice that click events happen on mouse down? I'm pretty confident that's a Bevy issue, though.

> * It would be nice have a new utility function in `UiRect` to set the `left/right` and `top/bottom` values together as it's probably my most used way of setting padding / margin when doing web development.
>   I could PR something like this if interested but understand if you want to keep the framework light.

Yeah I can see the usefulness of this. I wouldn't call it `from_x_y`, though. Maybe just `xy`?

connorgmeehan commented on examples/ui/toolbar.rs:
```diff
@@ -0,0 +1,257 @@
+        .add_system(toolbar_message_system)
+        .add_system(toolbar_style_system)
+        // EventWriter system after EventReaders to prevent ordering issues.
+        .add_system(button_interaction_system)
```
> I'm on Mac as well (M1 Max) and it runs fine for me. Although, I did notice that click events happen on mouse down? I'm pretty confident that's a Bevy issue, though.

I changed the ordering of the events and added some comments describing my reasoning and everything feels a lot faster.  It could have just been an order of execution / next frame issue (although I am still getting very slow responses in the `game_menu` example).

connorgmeehan commented:
Thank you for the code review! That was really helpful!  Especially refactoring the messy booleans into the `match` statement.

> I'm on Mac as well (M1 Max) and it runs fine for me. Although, I did notice that click events happen on mouse down? I'm pretty confident that's a Bevy issue, though.

I changed the ordering of the events and added some comments describing my reasoning and everything feels a lot faster. It could have just been an order of execution / next frame issue (although I am still getting very slow responses in the game_menu example).  This behavior feels a bit odd to me, do you know if this is the best practice?

```rust
    App::new()
        .add_plugins(DefaultPlugins)
        // Only run the app when there is user input. This will significantly reduce CPU/GPU use.
        .insert_resource(WinitSettings::desktop_app())
        .add_startup_system(setup)
        .add_state(ToolbarState::default())
        .add_event::<ToolbarEvent>()
        // EventReader systems added before event writer
        .add_system(toolbar_state_system)
        .add_system(toolbar_message_system)
        .add_system(toolbar_style_system)
        // EventWriter system after EventReaders to prevent ordering issues.
        .add_system(button_interaction_system)
        .run();
```
> Yeah I can see the usefulness of this. I wouldn't call it from_x_y, though. Maybe just xy?
Yeah `UiRect::xy(...)` sounds good to me.  I might open a PR later on.

Weibye commented:
❗ Please change the target branch from `bevyengine:latest` to `bevyengine:main` ❗ 

The `latest` branch points to whatever is the newest release. At the time of writing this is `0.9.0`. All development should target the `main` :)

Weibye reviewed (changes requested):
(Change target branch)

mockersf commented:
this was closed automatically during branch maintenance, not for the content
